src/Internal/BalanceTx/BalanceTx.purs

```diff
@@ -278,14 +269,29 @@ balanceTxWithConstraints transaction extraUtxos constraintsBuilder = do
 setTransactionCollateral :: Address -> Transaction -> BalanceTxM Transaction
 setTransactionCollateral changeAddr transaction = do
   nonSpendableSet <- asksConstraints _nonSpendableInputs
-  collateral <- do
-    rawCollateral <- liftEitherContract $ note CouldNotGetCollateral <$>
-      Wallet.getWalletCollateral
-    -- filter out UTxOs that are set as non-spendable in the balancer constraints
-    let
-      isSpendable = not <<< flip Set.member nonSpendableSet <<< _.input <<<
-        unwrap
-    pure $ Array.filter isSpendable rawCollateral
+  mbCollateralUtxos <- asksConstraints _collateralUtxos
+  -- We must filter out UTxOs that are set as non-spendable in the balancer
+  -- constraints
+  let
+    isSpendable = not <<< flip Set.member nonSpendableSet
+  collateral <- case mbCollateralUtxos of
+    -- if no collateral utxos are specified, use the wallet
+    Nothing -> Array.filter (isSpendable <<< _.input <<< unwrap) <$> do
+      liftEitherContract $ note CouldNotGetCollateral <$>
+        Wallet.getWalletCollateral
```

Comment:
I don't like that we select collateral before filtering out unspendable utxos. The collateral selection algorithm should be given a pre-filtered set of candidate utxos as in the `Just` branch below.

Reply:
We can't tell the wallet which utxos to not use